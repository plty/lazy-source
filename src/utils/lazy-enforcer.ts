import Thunk from '../interpreter/Thunk'
import { Value } from '../types'

export const unravel = (value: Thunk): Value => {
  // Used to check if there are any cyclic structures
  const unravelThunk = (v: Value | Thunk): Value => {
    if (v instanceof Thunk) {
      return v.value()
    } else if (Array.isArray(v)) {
      return v.map(x => unravelThunk(x))
    } else {
      return v
    }
  }

  return unravelThunk(value)
}