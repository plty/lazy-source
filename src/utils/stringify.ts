import { MAX_LIST_DISPLAY_LENGTH } from '../constants'
import Closure from '../interpreter/closure'
import { Value } from '../types'
import Thunk from '../interpreter/Thunk'

function makeIndent(indent: number | string): string {
  if (typeof indent === 'number') {
    if (indent > 10) {
      indent = 10
    }
    return ' '.repeat(indent)
  } else {
    if (indent.length > 10) {
      indent = indent.substring(0, 10)
    }
    return indent
  }
}

function indentify(indent: string, s: string): string {
  return s
    .split('\n')
    .map(v => indent + v)
    .join('\n')
}

export const stringify = function*(
  value: Thunk,
  indent: number | string = 2,
  splitlineThreshold = 80
): IterableIterator<Value> {
  // Used to check if there are any cyclic structures
  const ancestors = new Set()

  // Precompute useful strings
  const indentString = makeIndent(indent)
  const arrPrefix = '[' + indentString.substring(1)
  const objPrefix = '{' + indentString.substring(1)
  const arrSuffix = indentString.substring(0, indentString.length - 1) + ']'
  const objSuffix = indentString.substring(0, indentString.length - 1) + '}'

  // Util functions

  // Determines if the array/object containing these (stringified) values
  // should be multiline and indented or oneline
  const shouldMultiline = (valueStrs: string[]) =>
    indentString !== '' &&
    (valueStrs.join(', ').length > splitlineThreshold || valueStrs.some(s => s.includes('\n')))

  // Stringify functions
  // The real one is stringifyValue

  const stringifyArray = function*(xs: Value[], indentLevel: number) {
    ancestors.add(xs)
    const valueStrs = yield* concatGens(xs.map(x => () => stringifyValue(x, 0)))
    ancestors.delete(xs)

    if (shouldMultiline(valueStrs)) {
      if (xs.length === 2) {
        // It's (probably) a source list
        // Don't increase indent on second element
        // so long lists don't look like crap
        return `${arrPrefix}${indentify(
          indentString.repeat(indentLevel + 1),
          valueStrs[0]
        ).substring(indentString.length)},
${indentify(indentString.repeat(indentLevel), valueStrs[1])}${arrSuffix}`
      } else {
        // A regular array,
        // indent second element onwards to match with first element
        return `${arrPrefix}${indentify(
          indentString.repeat(indentLevel + 1),
          valueStrs.join(',\n')
        ).substring(indentString.length)}${arrSuffix}`
      }
    } else {
      return `[${valueStrs.join(', ')}]`
    }
  }

  const stringifyObject = function*(obj: object, indentLevel: number) {
    ancestors.add(obj)
    const keys: [string] = yield* concatGens(
      Object.entries(obj).map(entry => () => stringifyValue(Thunk.from(entry[0])), 0)
    )
    const vals: [string] = yield* concatGens(
      Object.entries(obj).map(entry => () => stringifyValue(Thunk.from(entry[1])), 0)
    )
    const valueStrs = zip(keys, vals).map(([key, val]) => {
      if (val.includes('\n')) {
        return key + ':\n' + indentify(indentString, val)
      } else {
        return key + ': ' + val
      }
    })
    ancestors.delete(obj)

    if (shouldMultiline(valueStrs)) {
      return `${objPrefix}${indentify(
        indentString.repeat(indentLevel + 1),
        valueStrs.join(',\n')
      ).substring(indentString.length)}${objSuffix}`
    } else {
      return `{${valueStrs.join(', ')}}`
    }
  }

  const stringifyValue = function*(thunk: Thunk, indentLevel: number = 0): IterableIterator<Value> {
    const v = yield* thunk.evaluate()
    if (v === null) {
      return 'null'
    } else if (v === undefined) {
      return 'undefined'
    } else if (ancestors.has(v)) {
      return '...<circular>'
    } else if (v instanceof Closure) {
      return v.toString()
    } else if (typeof v === 'string') {
      return JSON.stringify(v)
    } else if (typeof v !== 'object') {
      return v.toString()
    } else if (ancestors.size > MAX_LIST_DISPLAY_LENGTH) {
      return '...<truncated>'
    } else if (Array.isArray(v)) {
      return yield* stringifyArray(v, indentLevel)
    } else {
      return yield* stringifyObject(v, indentLevel)
    }
  }

  return yield* stringifyValue(value, 0)
}

function* concatGens(gens: (() => IterableIterator<Value>)[]): IterableIterator<Value> {
  return yield* gens.reduce(
    (acc, cur) => {
      return function*() {
        const rest = yield* acc()
        const v = yield* cur()
        return [...rest, v]
      }
    },
    function*() {
      return []
    }
  )()
}

function zip<T, U>(xs: T[], ys: U[]): [T, U][] {
  return xs.map((x, i) => [x, ys[i]])
}
