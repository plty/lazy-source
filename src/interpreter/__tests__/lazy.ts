import { expectDisplayResult, expectResult } from '../../utils/testing'
import { stripIndent } from '../../utils/formatters'

test('Numeric 5 resolves to 5 on lazy', () => {
  return expectResult(
    stripIndent`
    5;
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(5)
})

test('Plain pair works', () => {
  return expectResult(
    stripIndent`
      stringify(pair(1, 2));
    `,
    { chapter: 2, variant: 'lazy' }
  ).toBe('[1, 2]')
})

test('Plain list works', () => {
  return expectResult(
    stripIndent`
      stringify(list(1, 2, 3, 4, 5));
    `,
    { chapter: 2, variant: 'lazy' }
  ).toBe('[1, [2, [3, [4, [5, null]]]]]')
})

test('Infinite list works.', () => {
  return expectResult(
    stripIndent`
    const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
    const ones = pair(1, ones);
    stringify(take(ones, 5));
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe('[1, [1, [1, [1, [1, null]]]]]')
})

test('Infinite flip flops', () => {
  return expectResult(
    stripIndent`
    const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
    const flips = pair(0, flops);
    const flops = pair(1, flips);
    stringify(take(flips, 6));
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe('[0, [1, [0, [1, [0, [1, null]]]]]]')
})

test('Infinite prime list', () => {
  return expectResult(
    stripIndent`
    const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
    const sieve = (l) => pair(head(l), sieve(filter((x) => x % head(l) !== 0, l)));
    const nat = (n) => pair(n, nat(n+1));
    const primes = sieve(nat(2));
    stringify(take(primes, 5));
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe('[2, [3, [5, [7, [11, null]]]]]')
})

test('Ternary true ? 1 : head(null) resolves to 1 on lazy', () => {
  return expectResult(
    stripIndent`
    true ? 1 : head(null);
  `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(1)
})

test('No evaluation of unused variable', () => {
  return expectDisplayResult(
    stripIndent`
      const a = display(1);
      display(2);
    `,
    { chapter: 2, variant: 'lazy' }
  ).toStrictEqual(['2'])
})

test('No evaluation of unused argument', () => {
  return expectDisplayResult(
    stripIndent`
      const f = (a, b) => b + 5;
      const a = display(20);
      f(a, 10);
    `,
    { chapter: 2, variant: 'lazy' }
  ).toStrictEqual([])
})

test('No evaluation of unused argument', () => {
  return expectDisplayResult(
    stripIndent`
      const f = (a, b) => b + 5;
      const a = display(20);
      f(a + 3, 10);
    `,
    { chapter: 2, variant: 'lazy' }
  ).toStrictEqual([])
})

test('Delayed execution of display until required', () => {
  return expectDisplayResult(
    stripIndent`
      const a = display(1);
      const f = (a) => {
        display(2);
        display(a + 3);
      };
      f(a);
    `,
    { chapter: 2, variant: 'lazy' }
  ).toStrictEqual(['2', '1', '4'])
})

test('Delayed evaluation of operation until required', () => {
  return expectDisplayResult(
    stripIndent`
      const a = display(1);
      const f = (a) => {
        display(2);
        display(a + 3);
      };
      f(display(a + 4));
    `,
    { chapter: 2, variant: 'lazy' }
  ).toStrictEqual(['2', '1', '5', '8'])
})

test('Not stuck at resolving unused value', () => {
  return expectResult(
    stripIndent`
      const f = (x) => f(display(x));
      const a = f(0);
      display(2);
    `,
    { chapter: 2, variant: 'lazy' }
  ).toBe(2)
})
