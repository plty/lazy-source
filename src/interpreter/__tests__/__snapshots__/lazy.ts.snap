// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Infinite flip flops: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
      const flips = pair(0, flops);
const flops = pair(1, flips);
stringify(take(flips, 6));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": "[0, [1, [0, [1, [0, [1, null]]]]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Infinite prime list: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
      const sieve = (l) => pair(head(l), sieve(filter((x) => x % head(l) !== 0, l)));
const nat = (n) => pair(n, nat(n+1));
const primes = sieve(nat(2));
stringify(take(primes, 5));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": "[2, [3, [5, [7, [11, null]]]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Numeric 5 resolves to 5 on lazy: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
5;",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 5,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`Ternary true ? 1 : head(null) resolves to 1 on lazy: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
true ? 1 : head(null);",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": 1,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`const ones = pair(1, ones) resolves to an infinite list of ones.: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const take = (xs, n) => n === 0 || is_null(xs) ? null : pair(head(xs), take(tail(xs), n - 1));
      const ones = pair(1, ones);
stringify(take(ones, 5));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": "[1, [1, [1, [1, [1, null]]]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
